Source/JavaScriptCore/wasm/js/WasmToJS.cpp

```diff
@@ -419,7 +413,7 @@ Expected<MacroAssemblerCodeRef<WasmEntryPtrTag>, BindingFailure> wasmToJS(VM& vm
             break;
         }
         default:  {
-            if (Wasm::isFuncref(returnType) || Wasm::isExternref(returnType) || Wasm::isI31ref(returnType))
```

Comment:
I think we should make sure that we don't export internal WASM struct or array references until that part of the spec is ironed out. 

Otherwise, LGTM

Reply:
> I think we should make sure that we don't export internal WASM struct or array references until that part of the spec is ironed out.

This took a little bit more code, since it required adding additional error checking masm code in `WasmToJS.cpp` and `JSToWasm.cpp` (as you want the check at run-time, to avoid erroring when create the Wasm-JS stubs). I modeled the check after the old check that was used for disallowed I64 exports to JS.

The latest push has the new checks and some additional tests that exercise them. Also some minor fixes, like making `emitThrowWasmToJSException` explicitly pass the third argument to the operation call instead of implicitly assuming it's in the right register.